File: CBFS Filter Samples/Directory Hide/python/dirhide.py
# 
# CBFS Filter 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of CBFS Filter in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.callback.com/cbfsfilter
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from cbfsfilter import *

input = sys.hexversion<0x03000000 and raw_input or input

def ensureArg(args, prompt, index):
  if len(args) <= index:
    while len(args) <= index:
      args.append(None)
    args[index] = input(prompt)
  elif args[index] is None:
    args[index] = input(prompt)


import os

PROGRAM_NAME = "{713CC6CE-B3E2-4fd9-838D-E28F558F6866}"
FAKE_ALTITUDE_FOR_DEBUG = "360000.24"
ERROR_PRIVILEGE_NOT_HELD = 0x0522
ERROR_FILE_NOT_FOUND = 2
ERROR_PATH_NOT_FOUND = 3
ERROR_INVALID_NAME = 123

filter : CBFilter = None
folder_path_to_hide = ""
folder_name_to_hide = ""


# ---------------------------------
#      Helper functions
# ---------------------------------

def six_input(prompt):
    print(prompt)
    if sys.version_info[0] == 3:
        return input()
    #else:
    #    return raw_input(prompt)


def is_drive_letter(path_str):
    if not path_str:
        return False

    c = path_str[0]
    if c.isalpha() and len(path_str) == 2 and path_str[1] == ':':
        return True
    else:
        return False


def convert_relative_path_to_absolute(path_str):
    res = None
    if path_str and path_str.strip():
        res = path_str
        # Linux-specific case of using a home directory
        if path_str == "~" or path_str.startswith("~/"):
            home_dir = os.path.expanduser("~")
            if path_str == "~":
                return home_dir
            else:
                return os.path.join(home_dir, path_str[1:])
        elif not is_drive_letter(path_str):
            try:
                res = os.path.abspath(path_str)

                if res.startswith("\\\\") and not os.path.exists(res):
                    print(f"The network folder '{res}' does not exist.")
                elif not os.path.exists(res):
                    print(f"The path '{res}' does not exist.")
            except Exception as e:
                print(f"Error while converting to absolute path: {e}")
                res = None
    return res


# ---------------------------------
#        Event handlers
# ---------------------------------

def cb_after_enumerate_directory(params: CBFilterAfterEnumerateDirectoryEventParams):
    global filter
    global folder_name_to_hide

    # if the filesystem returns the name of the "hidden" directory, then block it
    if params.file_name == folder_name_to_hide:
        params.process_request = False
        process_name = filter.get_originator_process_name()
        print("[enumerate_directory]: %s (by %s) -> BLOCKED" % (params.file_name, process_name))


# ---------------------------------

def cb_before_create_file(params: CBFilterBeforeCreateFileEventParams):
    global filter
    global folder_name_to_hide, folder_path_to_hide

    file_name = params.file_name
    if platform.system() == "Windows":
        file_name = params.file_name.lower()

    # block attempts to create a new folder with the same name as the hidden folder
    if file_name == folder_path_to_hide:
        params.result_code = ERROR_INVALID_NAME
        process_name = filter.get_originator_process_name()
        print("[create_file]: %s (by %s) -> INVALID_NAME" % (params.file_name, process_name))
        return

    # block attempts to create any object in the hidden folder and its children
    folder = folder_path_to_hide + os.sep
    if file_name.startswith(folder):
        params.result_code = ERROR_PATH_NOT_FOUND
        process_name = filter.get_originator_process_name()
        print("[create_file]: %s (by %s) -> PATH_NOT_FOUND" % (params.file_name, process_name))


# ---------------------------------

def cb_before_open_file(params: CBFilterBeforeOpenFileEventParams):
    global filter, folder_name_to_hide, folder_path_to_hide

    file_name = params.file_name.lower()

    # block attempts to open the hidden folder itself
    if file_name == folder_path_to_hide:
        params.result_code = ERROR_FILE_NOT_FOUND
        process_name = filter.get_originator_process_name()
        print("[open_file]: %s (by %s) -> FILE_NOT_FOUND" % (params.file_name, process_name))
        return

    # block attempts to open any object in the hidden folder and its children
    folder = folder_path_to_hide + os.sep
    if file_name.startswith(folder):
        params.result_code = ERROR_PATH_NOT_FOUND
        process_name = filter.get_originator_process_name()
        print("[open_file]: %s (by %s) -> PATH_NOT_FOUND" % (params.file_name, process_name))


# ---------------------------------
#           Main code
# ---------------------------------

def banner():
    print("CBFS Filter Directory Hider Demo")
    print("The sample shows how to hide a directory from access by other programs.")


# ---------------------------------

def print_usage():
    print()
    print("Usage: %s <command> <parameters>" % os.path.basename(sys.argv[0]))
    print()
    print("Commands:")
    print("  i <cab_file>    - install the CBFilter driver from the CAB file")
    print("  u <cab_file>    - uninstall the CBFilter driver using the CAB file")
    print("  d               - check if the CBFilter driver is installed")
    print("  h <folder_name> - hide a folder")


# ---------------------------------

def check_driver():
    l_filter = CBFilter()
    state = l_filter.get_driver_status(PROGRAM_NAME)
    if state == MODULE_STATUS_NOT_PRESENT:
        print("CBFilter driver not installed")
    else:
        version = l_filter.get_driver_version(PROGRAM_NAME)
        ver_high, ver_low = version >> 32, version & 0xffffffff
        print("CBFilter driver version: %d.%d.%d.%d" % (
        ver_high >> 16, ver_high & 0xffff, ver_low >> 16, ver_low & 0xffff))
        if state != MODULE_STATUS_RUNNING:
            print("NOTE: the driver is NOT running; please restart the computer to let the driver start")


# ---------------------------------

def get_cab_file():  # -> str:
    if len(sys.argv) < 3:
        print("ERROR: No CAB file specified")
        print_usage()
        return ""

    cab_file = convert_relative_path_to_absolute(sys.argv[2])
    if cab_file is None:
        print("Error: Invalid Driver Path.")
        return
    if not os.path.isfile(cab_file):
        print("ERROR: CAB file not found")
        print("  ->", cab_file)
        return ""

    return cab_file


# ---------------------------------

def install():
    cab_file = get_cab_file()
    if cab_file == "":
        return

    print("Installing driver...")
    print("  ->", cab_file)
    print()

    l_filter = CBFilter()
    try:
        reboot_needed = l_filter.install(cab_file, PROGRAM_NAME, None, FAKE_ALTITUDE_FOR_DEBUG, INSTALL_REMOVE_OLD_VERSIONS, None)

        print("Driver installed successfully")
        version = l_filter.get_driver_version(PROGRAM_NAME)
        ver_high, ver_low = version >> 32, version & 0xffffffff
        print("  -> driver version: %d.%d.%d.%d" % (ver_high >> 16, ver_high & 0xffff, ver_low >> 16, ver_low & 0xffff))
        if reboot_needed:
            print("NOTE: Reboot the computer for the changes to take effect")
    except CBFSFilterError as err:
        if err.code == ERROR_PRIVILEGE_NOT_HELD:
            print("ERROR: administrator rights are required; please run the script as Administrator")
        else:
            print("ERROR: driver NOT installed")
            print("  -> (%d) %s" % (err.code, err.message))


# ---------------------------------

def uninstall():
    l_filter = CBFilter()

    state = l_filter.get_driver_status(PROGRAM_NAME)
    if state == MODULE_STATUS_NOT_PRESENT:
        print("Driver not installed")
        return

    cab_file = get_cab_file()
    if cab_file == "":
        return

    print("Uninstalling driver...")
    print("  ->", cab_file)
    print()

    try:
        reboot_needed = l_filter.uninstall(cab_file, PROGRAM_NAME, None, UNINSTALL_VERSION_ALL)

        print("Driver uninstalled successfully")
        if reboot_needed:
            print("NOTE: Reboot the computer for the changes to take effect")
    except CBFSFilterError as err:
        if err.code == ERROR_PRIVILEGE_NOT_HELD:
            print("ERROR: administrator rights are required; please run the script as Administrator")
        else:
            print("ERROR: failed to uninstall the driver")
            print("  -> (%d) %s" % (err.code, err.message))


# ---------------------------------

def hide():
    global filter, folder_path_to_hide, folder_name_to_hide

    if len(sys.argv) < 3:
        print("ERROR: No folder specified")
        print_usage()
        return ""

    folder_path_to_hide = convert_relative_path_to_absolute(os.path.normpath(sys.argv[2].lower()))
    if not os.path.exists(folder_path_to_hide):
        print("ERROR: folder does not exist - nothing to hide")
        print("  ->", folder_path_to_hide)
        return

    if not os.path.isdir(folder_path_to_hide):
        print("ERROR: the name doesn't seem to point to a folder")
        print("  ->", folder_path_to_hide)
        return

    # if the path ends with \ it's needed to remove it
    if folder_path_to_hide[-1] == '\\':
        folder_path_to_hide = folder_path_to_hide[:-1]

    parent = os.path.dirname(folder_path_to_hide)
    if parent == folder_path_to_hide:
        print("ERROR: cannot hide the root of a drive")
        print("  ->", parent)
        return

    folder_name_to_hide = os.path.basename(folder_path_to_hide)
    if platform.system() == "Windows":
        folder_name_to_hide = folder_name_to_hide.lower()

    filter = CBFilter()
    filter.initialize(PROGRAM_NAME)
    filter.on_after_enumerate_directory = cb_after_enumerate_directory
    filter.on_before_create_file = cb_before_create_file
    filter.on_before_open_file = cb_before_open_file

    print("Configuring the filter...")
    try:
        filter.add_filter_rule(parent, ACCESS_NONE,
                               FS_CE_AFTER_ENUMERATE_DIRECTORY,
                               FS_NE_NONE)
        print("  - filtered enumeration of folder:", parent)

        filter.add_filter_rule(folder_path_to_hide, ACCESS_NONE,
                               FS_CE_BEFORE_CREATE |
                               FS_CE_BEFORE_OPEN,
                               FS_NE_NONE)
        print("  - filtered opening and creation of folder:", folder_path_to_hide)

        file_mask = folder_path_to_hide + os.sep + "*.*"
        filter.add_filter_rule(file_mask, ACCESS_NONE,
                               FS_CE_BEFORE_CREATE |
                               FS_CE_BEFORE_OPEN,
                               FS_NE_NONE)
        print("  - filtered opening and creation of objects:", file_mask)

        filter.process_failed_requests = False
    except CBFSFilterError as err:
        print("ERROR: failed to configure the filter")
        print("  -> (%d) %s" % (err.code, err.message))
        return

    print("Configured")
    print("Starting the filter...")
    try:
        filter.start_filter(10000)
    except CBFSFilterError as err:
        print("ERROR: failed to start the filter")
        print("  -> (%d) %s" % (err.code, err.message))
        return

    print("The filter is working and the folder is hidden")
    print("  ->", folder_path_to_hide)

    while True:
        six_input("Press [Enter] to finish")
        try:
            print("Stopping the filter...")
            filter.delete_all_filter_rules()
            if filter.active:
                filter.stop_filter()
            print("Filter stopped")
            break
        except CBFSFilterError as err:
            print("ERROR: Failed to stop the filter")
            print("  -> (%d) %s" % (err.code, err.message))
            print("Will try again...")
            continue

    print("\nDemo finished")


# ---------------------------------

def test():
    folder = "C:\\Temp"

    print(os.path.basename(folder).lower())

    # pass


# ---------------------------------
#   Program body

banner()

if len(sys.argv) < 2:
    print_usage()
else:
    command = sys.argv[1]
    if command == "i":
        install()
    elif command == "u":
        uninstall()
    elif command == "d":
        check_driver()
    elif command == "h":
        hide()
    elif command == "t":
        test()
    else:
        print("ERROR: unknown command '%s'" % command)
        print_usage()

print()

File: CBFS Filter Samples/Directory Hide/c++/dirhide.cpp
/*
 * CBFS Filter 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of CBFS Filter in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.callback.com/cbfsfilter
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#define WIN32_LEAN_AND_MEAN        // Exclude rarely-used stuff from Windows headers

#define _CRT_SECURE_NO_DEPRECATE
#define _CRT_NON_CONFORMING_SWPRINTFS

// Windows Header Files:
#ifdef _WIN32
#include <windows.h>
#include <windowsx.h>
#endif

// C RunTime Header Files

#include <string>
#include <stdlib.h>
#include <stdio.h>
#include <conio.h>
#include <string.h>
#include <limits.h>
#include <malloc.h>
#include <memory.h>
#include <tchar.h>
#include <commctrl.h>
#include <assert.h>
#include <iostream>
#include <Shlwapi.h>
#include <filesystem>

#ifdef _UNICODE
#include "../../include/unicode/cbfsfilter.h"
#else
#include "../../include/cbfsfilter.h"
#endif

#ifdef UNIX
#include <unistd.h>
#endif

const TCHAR DirectorySeparatorChar =
#ifdef _WIN32
_T('\\');
#else
_T('/');
#endif

using namespace std;

#ifdef UNICODE
#define sout wcout
#define scin wcin
typedef std::wstring cbt_string;
#else
#define sout cout
#define scin cin
typedef std::string cbt_string;
#define _T(q) q
#endif

const TCHAR DirectorySeparator[] = { DirectorySeparatorChar, _T('\0') };

LPCTSTR g_Guid = _T("{713CC6CE-B3E2-4fd9-838D-E28F558F6866}");
BOOL g_Closing = FALSE;

void AddToLog(CBFilter* Filter, LPCTSTR Operation, LPCTSTR FileName, DWORD Status);

class CBFSFilter : public CBFilter
{
public:
    static TCHAR mPathToHide[MAX_PATH];
    static PTCHAR mOwnNameToHide;

    CBFSFilter() : CBFilter()
    {
    }

    INT FireFilterStop(CBFilterFilterStopEventParams* e)
    {
        return 0;
    }

    INT FireNotifyRenameOrMoveFile(CBFilterNotifyRenameOrMoveFileEventParams* e)
    {
        AddToLog(this, _T("NotifyRenameOrMoveFile"), e->FileName, NtStatusToWin32Error(e->Status));
        return 0;
    }

    INT FireBeforeCreateFile(CBFilterBeforeCreateFileEventParams* e)
    {
        if (!_tcsicmp(e->FileName, mPathToHide))
        {
            AddToLog(this, _T("BeforeCreateFile"), e->FileName, ERROR_FILE_NOT_FOUND);
            e->ResultCode = ERROR_FILE_NOT_FOUND;
            return 0;
        }

        const TCHAR* Separator = _tcsrchr(e->FileName, DirectorySeparatorChar);
        if (Separator && !_tcsnicmp(e->FileName, mPathToHide, Separator - e->FileName))
        {
            e->ResultCode = ERROR_PATH_NOT_FOUND;
        }
        /*AddToLog(this, _T("BeforeCreateFile"), e->FileName, NtStatusToWin32Error(e->Status));*/
        return 0;
    }

    INT FireBeforeOpenFile(CBFilterBeforeOpenFileEventParams* e)
    {
        if (!_tcsicmp(e->FileName, mPathToHide))
        {
            AddToLog(this, _T("BeforeOpenFile"), e->FileName, ERROR_FILE_NOT_FOUND);
            e->ResultCode = ERROR_FILE_NOT_FOUND;
            return 0;
        }

        const TCHAR* Separator = _tcsrchr(e->FileName, DirectorySeparatorChar);
        if (Separator && !_tcsnicmp(e->FileName, mPathToHide, Separator - e->FileName))
        {
            e->ResultCode = ERROR_PATH_NOT_FOUND;
        }
        //AddToLog(this, _T("NotifyOpenFile"), e->FileName, NtStatusToWin32Error(e->Status));
        return 0;
    }

    INT FireAfterEnumerateDirectory(CBFilterAfterEnumerateDirectoryEventParams* e)
    {
        // if the filesystem returned the name of the directory that we aim to hide, then hide it
        if (!_tcsicmp(e->FileName, mOwnNameToHide))
        {
            AddToLog(this, _T("AfterEnumerateDirectory"), e->FileName, ERROR_FILE_NOT_FOUND);
            e->ProcessRequest = FALSE;
            return 0;
        }

        //AddToLog(this, _T("NotifyEnumerateDirectory"), buf, NtStatusToWin32Error(e->Status));
        return 0;
    }
};

TCHAR CBFSFilter::mPathToHide[MAX_PATH] = { 0 };
PTCHAR CBFSFilter::mOwnNameToHide = NULL;

CBFSFilter g_CBFSFlt;

#define CBFSFILTER_LOG_OPERATION             0
#define CBFSFILTER_LOG_PATH                  1
#define CBFSFILTER_LOG_ORIGINATOR_PROCESS    2
#define CBFSFILTER_LOG_PROCESS_ID            3
#define CBFSFILTER_LOG_USER_NAME             4
#define CBFSFILTER_LOG_RESULT                5

CRITICAL_SECTION    g_LogListLock = { 0, };

typedef __int64 int64;

const cbt_string ALTITUDE_FAKE_VALUE_FOR_DEBUG(TEXT("360000.24"));

const cbt_string product_id(TEXT("{713CC6CE-B3E2-4fd9-838D-E28F558F6866}"));
const cbt_string strInvalidOption(TEXT("Invalid option \"%s\"\n"));

const std::string strFilterError("Error %d (%s)\n");

void HideDir(LPTSTR PathToHide);
void UnhideDir();

void SetFilter(LPCTSTR PathToMonitor, LPCTSTR FilenameMask);
void DeleteFilter();

using namespace std;

void Banner(void)
{
    printf("CBFS Filter DirHide Demo Copyright (c) 2017-2023, Callback Technologies, Inc.\n\n");
}

void Usage(void)
{
    printf("Usage: dirhide [-<switch 1> ... -<switch N>] <path to hide>\n\n");
    printf("<Switches>\n");
    printf("  -drv <path to cbfilter.cab> - Install the driver from the specified CAB file\n");
    printf("  -- Stop switches scanning\n");
}

// ----------------------------------------------------------------------------------

int CheckResult(int result)
{
    if (result != 0)
    {
        fprintf(stderr, strFilterError.c_str(), g_CBFSFlt.GetLastErrorCode(), g_CBFSFlt.GetLastError());
    }
    return result;
}

void CheckDriver()
{
    int state, number;

    state = g_CBFSFlt.GetDriverStatus(product_id.c_str());
    if (CheckResult(g_CBFSFlt.GetLastErrorCode()))
        exit(0);
    if (state == cbfConstants::MODULE_STATUS_RUNNING)
    {
        int64 version;
        version = g_CBFSFlt.GetDriverVersion(product_id.c_str());
        if (CheckResult(g_CBFSFlt.GetLastErrorCode()))
            exit(0);
        printf("The driver is installed, version: %d.%d",
            (int)((version & 0x7FFF000000000000) >> 48),
            (int)((version & 0xFFFF00000000) >> 32));
        number = (int)((version & 0xFFFF0000) >> 16);
        if (number != 0)
            printf(".%d", number);
        number = (int)(version & 0xFFFF);
        if (number != 0)
            printf(".%d", number);
        printf("\n");
    }
    else
    {
        printf("Error: the driver is not installed\n");
        exit(0);
    }
}

int optcmp(cbt_string arg, string opt)
{
    int i = 0;
    while (1)
    {
        if (arg[i] >= 'A' && arg[i] <= 'Z')
            arg[i] = arg[i] - 'A' + 'a';
        if (arg[i] != opt[i])
            return 0;
        if (arg[i] == 0)
            return 1;
        i++;
    }
}

bool IsDriveLetter(const cbt_string& path) {
    if (path.empty())
        return false;

    wchar_t c = path[0];
    if (((c >= L'A' && c <= L'Z') || (c >= L'a' && c <= L'z')) && path.size() == 2 && path[1] == L':')
        return true;
    else
        return false;
}

bool IsAbsolutePath(const cbt_string& path) {
    if (path.empty()) {
        return false;
    }

#ifdef _WIN32
    // On Windows, check if the path starts with a drive letter followed by a colon and a separator
    if ((path.size() >= 3) && iswalpha(path[0]) && (path[1] == L':') && (path[2] == L'\\' || path[2] == L'/')) {
        return true;
    }

    // Check for UNC paths (e.g., \\server\share)
    if (path.size() >= 2 && (path[0] == L'\\') && (path[1] == L'\\')) {
        return true;
    }
#else
    // On Linux and Unix, check if the path starts with a '/'
    if (path[0] == _T('/')) {
        return true;
    }
#endif

    return false;
}

cbt_string ConvertRelativePathToAbsolute(const cbt_string& path, bool acceptDriveLetter = false) {
    cbt_string res;

    if (!path.empty()) {
        res = path;

#ifndef _WIN32
        // Linux/Unix-specific case of using a home directory
        if (path == "~" || path.find("~/") == 0) {
            const char* homeDir = getenv("HOME");

            if (path == "~") {
                return homeDir ? homeDir : "";
            }
            else {
                return homeDir ? cbt_string(homeDir) + path.substr(1) : "";
            }
        }
#endif
        if (!IsAbsolutePath(path)) {
#ifdef _WIN32
            if (IsDriveLetter(path)) {
                if (!acceptDriveLetter) {
                    sout << L"The path '" << path << L"' cannot be equal to the drive letter" << std::endl;
                    return _T("");
                }
                return res;
            }
            wchar_t currentDir[_MAX_PATH];
            const char pathSeparator = '\\';
            if (_wgetcwd(currentDir, _MAX_PATH) == nullptr) {
                sout << "Error getting current directory." << std::endl;
                return _T("");
            }
#else
            char currentDir[PATH_MAX];
            const char pathSeparator = '/';
            if (getcwd(currentDir, sizeof(currentDir)) == nullptr) {
                sout << "Error getting current directory." << std::endl;
                return _T("");
            }
#endif
            cbt_string currentDirStr(currentDir);

            // Ensure that the current directory has a trailing backslash
            if (currentDirStr.back() != pathSeparator) {
                currentDirStr += pathSeparator;
            }

            return currentDirStr + path;
        }
    }
    else {
        sout << L"Error: The input path is empty." << std::endl;
        return _T("");
    }
    return path;
}

int wmain(int argc, wchar_t* argv[]) {
    INT drv_reboot = 0;
    LPTSTR PathToHide = NULL;
    int res_code = 0, argi = 0, arg_len = 0, stop_opt = 0;

    Banner();

    // if no parameters specified, show usage and quit
    if (argc < 2) {
        Usage();
        CheckDriver();
        return 0;
    }

    for (argi = 1; argi < argc; argi++) {
        cbt_string arg = argv[argi];
        arg_len = (int)arg.length();
        if (arg_len > 0) {
            if ((arg[0] == '-') && !stop_opt) {
                if (optcmp(arg, "-drv")) {
                    arg = argv[++argi];
                    if (argi < argc) {
                        _tprintf(_T("Installing the driver from '%s'\n"), arg.c_str());
                        cbt_string drv_path_wstr = ConvertRelativePathToAbsolute(arg);
                        if (drv_path_wstr.empty()) {
                            printf("Error: Invalid Driver Path\n");
                            exit(1); 
                        }
                        drv_reboot = g_CBFSFlt.Install(const_cast<LPTSTR>(drv_path_wstr.c_str()), product_id.c_str(), NULL,
                            ALTITUDE_FAKE_VALUE_FOR_DEBUG.c_str(), 0, NULL);// cbfConstants::INSTALL_REMOVE_OLD_VERSIONS);

                        int errcode = g_CBFSFlt.GetLastErrorCode();
                        if (errcode != 0) {
                            if (errcode == ERROR_PRIVILEGE_NOT_HELD)
                                printf("The driver could not be installed, please try to install with administrator rights\n");
                            else
                                printf("The driver could not be installed, the error is %d (%s)\n", errcode, g_CBFSFlt.GetLastError());

                            exit(errcode);
                        }
                        printf("Drivers installed successfully\n");
                        if (drv_reboot != 0) {
                            printf(", reboot is required\n");
                            exit(1);
                        }
                    }
                }
                else
                    fwprintf(stderr, strInvalidOption.c_str(), arg.c_str());
            }
            else {
                cbt_string path_to_hide_wstr = ConvertRelativePathToAbsolute(_tcsdup(arg.c_str()));
                if (path_to_hide_wstr.empty()) {
                    printf("Error: Invalid Path To Hide\n");
                    exit(1); 
                }
                PathToHide = _tcsdup(path_to_hide_wstr.c_str());
                DWORD attr = GetFileAttributes(PathToHide);
                if (attr == INVALID_FILE_ATTRIBUTES || (attr & FILE_ATTRIBUTE_DIRECTORY) == 0) {
                    _tprintf(_T("ERROR: the specified path '%s' does not point to an existing directory\n"), PathToHide);
                    exit(3);
                }
            }
        }
    }

    CheckDriver();

    // Probably, we have just installed the driver, so we can quit without monitoring anything
    if (PathToHide == NULL)
        exit(0);

    _tprintf(_T("Press any key to start hiding, then press any key to stop and quit.\n"));

    _getch();

    InitializeCriticalSection(&g_LogListLock);
    try {
        HideDir(PathToHide);
        //SetFilter(PathToMonitor, FilenameMask);
        _getch();
        UnhideDir();
        //DeleteFilter();
    }
    catch (...) {
    }

    if (PathToHide)
        free(PathToHide);

    return 0;
}

void AddToLog(CBFilter* Filter, LPCTSTR Operation, LPCTSTR FileName, DWORD Status)
{
    EnterCriticalSection(&g_LogListLock);
    _tprintf(_T("%s - %s (operation status = %d)\n"), Operation, FileName, Status);
    LeaveCriticalSection(&g_LogListLock);
}

void HideDir(LPTSTR PathToHide)
{
    TCHAR Path[MAX_PATH], Parent[MAX_PATH];

    _tcscpy(CBFSFilter::mPathToHide, PathToHide);

    // strip ending backslashes
    while (CBFSFilter::mPathToHide[_tcslen(CBFSFilter::mPathToHide) - 1] == DirectorySeparatorChar)
    {
        CBFSFilter::mPathToHide[_tcslen(CBFSFilter::mPathToHide) - 1] = _T('\0');
    }

    CBFSFilter::mOwnNameToHide = _tcsrchr(CBFSFilter::mPathToHide, DirectorySeparatorChar);
    // Check that it's not a root to be hidden
    if (!CBFSFilter::mOwnNameToHide)
    {
        _tprintf(_T("ERROR: The specified directory is root, and roots cannot be hidden."));
        return;
    }

    _tcsncpy(Parent, CBFSFilter::mPathToHide, CBFSFilter::mOwnNameToHide - CBFSFilter::mPathToHide + 1);
    Parent[CBFSFilter::mOwnNameToHide - CBFSFilter::mPathToHide + 1] = 0;//_tcsncpy: No null wide character is implicitly appended at the end of destination if source is longer than num (thus, in this case, destination may not be a null terminated C wide string).

    CBFSFilter::mOwnNameToHide++;

    // Hide directory from the parent's enumerations
    g_CBFSFlt.AddFilterRule(Parent,
        cbfConstants::ACCESS_NONE,
        cbfConstants::FS_CE_AFTER_ENUMERATE_DIRECTORY,
        cbfConstants::FS_NE_NONE);

    // Prevent direct opening of the directory
    g_CBFSFlt.AddFilterRule(CBFSFilter::mPathToHide,
        cbfConstants::ACCESS_NONE,
        cbfConstants::FS_CE_BEFORE_OPEN |
        cbfConstants::FS_CE_BEFORE_CREATE,
        cbfConstants::FS_NE_NONE);

    // Prevent direct operations on files in the directory
    _tcscpy(Path, CBFSFilter::mPathToHide);
    _tcscat(Path, DirectorySeparator);
    _tcscat(Path, _T("*.*"));
    g_CBFSFlt.AddFilterRule(Path,
        cbfConstants::ACCESS_NONE,
        cbfConstants::FS_CE_BEFORE_OPEN |
        cbfConstants::FS_CE_BEFORE_CREATE,
        cbfConstants::FS_NE_NONE);

    int retVal = g_CBFSFlt.SetProcessFailedRequests(TRUE);
    if (0 == retVal)
        retVal = g_CBFSFlt.Initialize(g_Guid);
    if (0 == retVal)
        retVal = g_CBFSFlt.StartFilter(30000);

    if (0 != retVal)
        CheckResult(retVal);
}

void UnhideDir()
{
    g_CBFSFlt.DeleteAllFilterRules();
    g_CBFSFlt.StopFilter();
}


